Examples/Io/Performance/ACTFW/Io/Performance/CKFPerformanceWriter.cpp

```diff
@@ -0,0 +1,188 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include <numeric>
+#include <stdexcept>
+
+#include <TFile.h>
+#include <TTree.h>
+
+#include "ACTFW/EventData/SimParticle.hpp"
+#include "ACTFW/Io/Performance/CKFPerformanceWriter.hpp"
+#include "ACTFW/Utilities/Paths.hpp"
+#include "Acts/EventData/MultiTrajectoryHelpers.hpp"
+//#include "Acts/Utilities/Helpers.hpp"
+
+FW::CKFPerformanceWriter::CKFPerformanceWriter(
+    FW::CKFPerformanceWriter::Config cfg, Acts::Logging::Level lvl)
+    : WriterT(cfg.inputTrajectories, "CKFPerformanceWriter", lvl),
+      m_cfg(std::move(cfg)),
+      m_effPlotTool(m_cfg.effPlotToolConfig, lvl),
+      m_fakeRatePlotTool(m_cfg.fakeRatePlotToolConfig, lvl),
+      m_trackSummaryPlotTool(m_cfg.trackSummaryPlotToolConfig, lvl) {
+  // Input track and truth collection name
+  if (m_cfg.inputTrajectories.empty()) {
+    throw std::invalid_argument("Missing input trajectories collection");
+  }
+  if (m_cfg.inputParticles.empty()) {
+    throw std::invalid_argument("Missing input particles collection");
+  }
+  if (m_cfg.outputFilename.empty()) {
+    throw std::invalid_argument("Missing output filename");
+  }
+
+  // the output file can not be given externally since TFile accesses to the
+  // same file from multiple threads are unsafe.
+  // must always be opened internally
+  auto path = joinPaths(m_cfg.outputDir, m_cfg.outputFilename);
+  m_outputFile = TFile::Open(path.c_str(), "RECREATE");
+  if (not m_outputFile) {
+    throw std::invalid_argument("Could not open '" + path + "'");
+  }
+
+  // initialize the plot tools
+  m_effPlotTool.book(m_effPlotCache);
+  m_fakeRatePlotTool.book(m_fakeRatePlotCache);
+  m_trackSummaryPlotTool.book(m_trackSummaryPlotCache);
+}
+
+FW::CKFPerformanceWriter::~CKFPerformanceWriter() {
+  m_effPlotTool.clear(m_effPlotCache);
+  m_fakeRatePlotTool.clear(m_fakeRatePlotCache);
+  m_trackSummaryPlotTool.clear(m_trackSummaryPlotCache);
+  if (m_outputFile) {
+    m_outputFile->Close();
+  }
+}
+
+FW::ProcessCode FW::CKFPerformanceWriter::endRun() {
+  if (m_outputFile) {
+    m_outputFile->cd();
+    m_effPlotTool.write(m_effPlotCache);
+    m_fakeRatePlotTool.write(m_fakeRatePlotCache);
+    m_trackSummaryPlotTool.write(m_trackSummaryPlotCache);
+    ACTS_INFO("Wrote performance plots to '" << m_outputFile->GetPath() << "'");
```

Comment:
Does `m_outputFIle` need to be closed? I'm never sure..

Reply:
I think so. It's has closed in the destructor.